numba/targets/linalg.py

```diff
@@ -1004,3 +1004,103 @@ def svd_impl(a, full_matrices=1):
             return (u.T, s, vt.T)
 
         return svd_impl
+
+    @overload(numpy.linalg.qr)
+    def qr_impl(a):
+        ensure_lapack()
+
+        _check_linalg_matrix(a, "qr")
+
+        # Need two functions, the first computes R, storing it in the upper
+        # triangle of A with the below diagonal part of A containing elementary
+        # reflectors needed to construct Q. The second turns the below diagonal
+        # entries of A into Q, storing Q in A (creates orthonormal columns from
+        # the elementary reflectors).
+
+        numba_ez_geqrf_sig = types.intc(
+            types.char,  # kind
+            types.intp,  # m
+            types.intp,  # n
+            types.CPointer(a.dtype),  # a
+            types.intp,  # lda
+            types.CPointer(a.dtype),  # tau
+        )
+
+        numba_ez_geqrf = types.ExternalFunction("numba_ez_geqrf",
+                                                numba_ez_geqrf_sig)
+
+        numba_ez_xxgqr_sig = types.intc(
+            types.char,  # kind
+            types.intp,  # m
+            types.intp,  # n
+            types.intp,  # k
+            types.CPointer(a.dtype),  # a
+            types.intp,  # lda
+            types.CPointer(a.dtype),  # tau
+        )
+
+        numba_ez_xxgqr = types.ExternalFunction("numba_ez_xxgqr",
+                                                numba_ez_xxgqr_sig)
+
+        kind = ord(get_blas_kind(a.dtype, "qr"))
+
+        F_layout = a.layout == 'F'
+
+        def qr_impl(a):
+            n = a.shape[-1]
+            m = a.shape[-2]
+
+            _check_finite_matrix(a)
+
+            # copy A as it will be destroyed
+            if F_layout:
+                q = numpy.copy(a)
+            else:
+                q = numpy.asfortranarray(a)
+
+            lda = m
+
+            minmn = min(m, n)
+            tau = numpy.empty((minmn), dtype=a.dtype)
+
+            ret = numba_ez_geqrf(
+                kind,  # kind
+                m,  # m
+                n,  # n
+                q.ctypes,  # a
+                m,  # lda
+                tau.ctypes  # tau
+            )
+            if ret < 0:
+                fatal_error_func()
+                assert 0   # unreachable
+
+            # pull out R, there is undoubtedly a more elegant way
+            # this is transposed in memory because Fortran
+            r = numpy.zeros((minmn, n), dtype=a.dtype)
+
+            for i in range(minmn):
+                for j in range(i, n):
+                    r[i, j] = q[i, j]
```

Comment:
I'll edit my conclusion :) The slice on the `Q` array leads to an `F contig` array being returned. I've therefore modified the code so that `R` is allocated with switched dimensions and then used transposed and hence is `F contig`. This is beneficial in that when extracting `R` (in the decomposition phase) from what will become `Q`, the arrays are both `F contig` and are walked in Fortran order, which is more efficient. I've also split the extraction loop to reduce comparisons on the upper limit of the vector lengths when copying in the triangle and possible square parts. The end result of all this being that `Q` and `R` are returned `F contig` and walked the same way! Patch 0e297e3 adds in the test you describe above, I've applied this test to all the `np.linalg` functions.